crates/miden-tx/src/host/mod.rs

```diff
@@ -506,23 +524,140 @@ where
 
-        Ok(())
+            if sn_script_data.len() != 8 {
+                return Err(TransactionKernelError::MalformedRecipientData(
+                    sn_script_data.to_vec(),
+                ));
+            }
+
+            let sn_hash = Word::new([
+                sn_script_data[0],
+                sn_script_data[1],
+                sn_script_data[2],
+                sn_script_data[3],
+            ]);
+            let script_root = Word::new([
+                sn_script_data[4],
+                sn_script_data[5],
+                sn_script_data[6],
+                sn_script_data[7],
+            ]);
+
+            let sn_hash_data = process
+                .advice_provider()
+                .get_mapped_values(&sn_hash)
+                .ok_or_else(|| TransactionKernelError::MalformedRecipientData(vec![]))?;
+
+            if sn_hash_data.len() != 8 {
+                return Err(TransactionKernelError::MalformedRecipientData(sn_hash_data.to_vec()));
+            }
+
+            let serial_num =
+                Word::new([sn_hash_data[0], sn_hash_data[1], sn_hash_data[2], sn_hash_data[3]]);
+
+            let num_inputs = process
+                .advice_provider()
+                .get_mapped_values(&inputs_commitment)
+                .and_then(|inputs| inputs.iter().rposition(|&x| x != ZERO).map(|pos| pos + 1))
+                .unwrap_or(0);
+
+            let script_data = process.advice_provider().get_mapped_values(&script_root);
+
+            if script_data.is_none() {
+                let inputs = self.extract_note_inputs(process, &inputs_commitment, num_inputs)?;
+
+                return Ok(TransactionEventHandling::Unhandled(TransactionEventData::NoteScript {
+                    script_root,
+                    metadata,
+                    recipient_digest,
+                    note_idx,
+                    note_inputs: inputs,
+                    serial_num,
+                }));
+            }
+
+            let inputs = self.extract_note_inputs(process, &inputs_commitment, num_inputs)?;
+
+            let script = NoteScript::try_from(script_data.unwrap()).map_err(|source| {
+                TransactionKernelError::MalformedNoteScript {
+                    data: script_data.unwrap().to_vec(),
+                    source,
+                }
+            })?;
+
+            Some(NoteRecipient::new(serial_num, script, inputs))
+        } else {
+            None
+        };
+
+        let note_builder = OutputNoteBuilder::new(metadata, recipient_digest, recipient)?;
+        self.insert_output_note_builder(note_idx, note_builder)?;
+
+        Ok(TransactionEventHandling::Handled(Vec::new()))
+    }
+
+    /// Extracts and validates note inputs from the advice provider.
+    fn extract_note_inputs(
+        &self,
+        process: &ProcessState,
+        inputs_commitment: &Word,
+        num_inputs: usize,
+    ) -> Result<NoteInputs, TransactionKernelError> {
```

Comment:
I would remove `note_inputs` from the parameters to this method, and instead, would try to determine them here. The way it could work:

1. Try to determine `num_inputs` based on the last non-zero element (same as you have in the code above).
2. Then, build `NoteInputs` and check if the hash matches `inputs_commitment`.
3. If not, increment `num_inputs` by one, and try again.
4. Keep doing this for potentially 6 more times.
5. If `num_inputs` grows to the size of `inputs_data` and there is still no match, return an error.

This is a bit hacky and maybe there are better ways of doing this - but nothing came immediately to mind.

---------

Also nit: I don't think we use `self` in this function - so, it could probably be a "free" function.